### PR TITLE
Added GitHub Actions CI, Refactoring

### PR DESCRIPTION
**Pull request by lobis**
Large lobis  [](https://gitlab.cern.ch/rest-for-physics/restg4/-/commits/master) [Powered by Pull Request Badge](https://pullrequestbadge.com/?utm_medium=github&utm_source=rest-for-physics&utm_campaign=badge_info)

I have added GitHub actions support, still not fully replicating the pipeline, but this will be done soon.

I have also done some refactoring, better ordering imports and removed old comments which I believe were from copy-pasting the Geant4 examples.

### CODE REVIEW

**Review by jgalan — APPROVED** [review]
Perhaps it would be better to place the validation.yml into a different pull-request?

Header and reformatting seems ok. I guess some of those changes could be maintained by `clang-format` job being run at the pipeline, as for other repositories.

File: .github/workflows/validation.yml [hunk 1]
```diff
@@ -0,0 +1,62 @@
...
+
+defaults:
+  run:
+    shell: bash -ieo pipefail {0}
+
+jobs:
+  build:
+    runs-on: ubuntu-latest
+    container:
+      image: lobis/root-geant4-garfieldpp:cxx14_ROOTv6-25-01_Geant4v10.4.3
```
Comment (juanangp) [inline .github/workflows/validation.yml:21]: I am not an expert on this topic, but shall the image point to lobis?

File: .github/workflows/validation.yml [hunk 2]
```diff
@@ -0,0 +1,62 @@
...
+
+    strategy:
+      matrix:
+        framework-branch: [master, development]
+
+    steps:
+      - name: Source ROOT and Geant4
+        run: |
+          source /root/.bashrc
```
Comment (juanangp) [inline .github/workflows/validation.yml:30]: There is no better way to do it?

File: .github/workflows/validation.yml (hunk 1 above)
Comment (jgalan) [reply .github/workflows/validation.yml:21]: Thats the image we are using for validation, created by @lobis. Why that comment? What do you mean? To have a docker account named `rest-for-physics`?

Comment (lobis) [reply .github/workflows/validation.yml:21]: Yes, that would be better I guess. I guess if we do that, we should host the docker image via github packages, which is the github way of doing this.

File: include/RunAction.hh [hunk 3]
```diff
@@ -1,53 +1,18 @@
...
-#include "G4UserRunAction.hh"
-#include "globals.hh"
-
+#include <G4UserRunAction.hh>
 #include <fstream>
+#include <globals.hh>
 #include <iostream>
+#include <map>
+
 using namespace std;
```
Comment (juanangp) [inline include/RunAction.hh:11]: We should remove namespaces from the headers, perhaps not in this MR, but create an issue about that

File: .github/workflows/validation.yml (hunk 2 above)
Comment (lobis) [reply .github/workflows/validation.yml:30]: Not that I found, seems like declaring the environment variables is not enough (or I am not declaring all of them). So far this works, but if you can think of a better way let me know.

File: include/RunAction.hh (hunk 3 above)
Comment (lobis) [reply include/RunAction.hh:11]: Agree

File: src/DetectorConstruction.cc [hunk 4]
```diff
@@ -31,18 +29,12 @@ G4VPhysicalVolume* DetectorConstruction::Construct() {
...
     TString geometryFile = restG4Metadata->Get_GDML_Filename();
 
-    // char originDirectory[255];
-    // sprintf( originDirectory, "%s", get_current_dir_name() );
-
-    // char buffer[255];
-    // sprintf( buffer, "%s", (char *) restG4Metadata->GetGeometryPath().Data() );
-    // chdir( buffer );
     char originDirectory[256];
     sprintf(originDirectory, "%s", getenv("PWD"));
```
Comment (juanangp) [inline src/DetectorConstruction.cc:33]: The use of environmental variables in the code is not a good practice in general.
In this particular case you can use "std::filesystem::current_path()".

Comment (lobis) [reply src/DetectorConstruction.cc:33]: That is only a feature of C++17 if I'm correct, sadly

File: .github/workflows/validation.yml (hunk 2 above)
Comment (juanangp) [reply .github/workflows/validation.yml:30]: I don't have much experience on this topic, but I can have a look...

File: src/DetectorConstruction.cc (hunk 4 above)
Comment (juanangp) [reply src/DetectorConstruction.cc:33]: Indeed, another reason to move to C++17

**Comment by lobis** [comment]
> Perhaps it would be better to place the validation.yml into a different pull-request?
> 
> Header and reformatting seems ok. I guess some of those changes could be maintained by `clang-format` job being run at the pipeline, as for other repositories.

Yes, maybe it's not a good idea to mix things in the pull request, but since it won't alter functionality and it is very easy to separate I thought it was OK

**Comment by jgalan** [comment]
Right it is ok, was just to do not exceed the limit of 200 additions.

**Pull request merged**